### PR TITLE
Deprecate ``doc`` attribute

### LINKED ISSUE #1414: Deprecate `doc` attribute?
With the new addition of `doc_node` (#1276), should the "old" `doc` attribute be deprecated and later removed?

/CC: @DanielNoord

### PR DESCRIPTION
**Pull request by DanielNoord**
## Steps

- [x] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [x] Write a good description on what the PR does.

## Description

Closes #1414.

/CC @cdce8p I just saw I took a completely different approach than what you had in mind. Is this something we should pursue or do you want to use your method? (I didn't look completely in to how that works)

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |
| ✓   | :hammer: Refactoring   |

### CODE REVIEW

File: astroid/raw_building.py [hunk 1]
```diff
@@ -119,6 +119,13 @@ def build_module(name: str, doc: Optional[str] = None) -> nodes.Module:
 def build_class(name, basenames=(), doc=None):
     """create and initialize an astroid ClassDef node"""
     node = nodes.ClassDef(name, doc)
```
Comment (Pierre-Sassoulas) [inline astroid/raw_building.py:121]: I don't understand why we have to use the postinit here ?

File: astroid/brain/brain_builtin_inference.py [hunk 2]
```diff
@@ -570,12 +570,16 @@ def infer_property(node, context=None):
...
+        doc=getattr(inferred, "doc_node.value", None),
         lineno=node.lineno,
         parent=node,
         col_offset=node.col_offset,
     )
-    prop_func.postinit(body=[], args=inferred.args)
+    prop_func.postinit(
+        body=[],
+        args=inferred.args,
+        doc_node=getattr(inferred, "doc_node", None),
```
Comment (Pierre-Sassoulas) [inline astroid/brain/brain_builtin_inference.py:581]: Isn't it set line 573 in the constructor of Property ?

Comment (DanielNoord) [reply astroid/brain/brain_builtin_inference.py:581]: That's the `doc` for `Property.doc`, this is for `doc_node`. I don't know why exactly but not using `getattr` crashes here, so we need to use `getattr`.

File: astroid/brain/brain_functools.py [hunk 3]
```diff
@@ -101,7 +101,9 @@ def _functools_partial_inference(node, context=None):
     partial_function = objects.PartialFunction(
         call,
         name=inferred_wrapped_function.name,
-        doc=inferred_wrapped_function.doc,
+        doc=inferred_wrapped_function.doc_node.value
```
Comment (cdce8p) [inline astroid/brain/brain_functools.py:104]: It seems like both `PartialFunction` and `Property` should accept `doc_node` instead.

File: astroid/interpreter/objectmodel.py [hunk 4]
```diff
@@ -584,7 +598,10 @@ def attr___name__(self):
     @property
     def attr___doc__(self):
         return node_classes.Const(
-            value=self._instance.parent.doc, parent=self._instance
+            value=self._instance.parent.doc_node.value
+            if self._instance.parent.doc_node
+            else None,
```
Comment (cdce8p) [inline astroid/interpreter/objectmodel.py:603]:
```suggestion
            value=getattr(self._instance.parent.doc_node, "value", None),
```
Not the best solution but a bit less ugly.

File: astroid/nodes/as_string.py [hunk 5]
```diff
@@ -183,7 +183,7 @@ def visit_classdef(self, node):
             args.append("metaclass=" + node._metaclass.accept(self))
         args += [n.accept(self) for n in node.keywords]
         args = f"({', '.join(args)})" if args else ""
-        docs = self._docs_dedent(node.doc) if node.doc else ""
+        docs = self._docs_dedent(node.doc_node.value) if node.doc_node else ""
```
Comment (cdce8p) [inline astroid/nodes/as_string.py:186]: Should `self._docs_dedent` accept `nodes.Const | None` instead?

File: astroid/nodes/scoped_nodes/scoped_nodes.py [hunk 6]
```diff
@@ -457,8 +457,8 @@ def __init__(
         self.name = name
         """The name of the module."""
 
-        self.doc = doc
-        """The module docstring."""
+        # TODO: Remove this attribute
+        self._doc = doc
```
Comment (cdce8p) [inline astroid/nodes/scoped_nodes/scoped_nodes.py:461]: At least for `Module.__init__` `doc` is still a required argument. I think we can add a default value.

File: astroid/nodes/scoped_nodes/scoped_nodes.py [hunk 7]
```diff
@@ -513,6 +513,16 @@ def _get_stream(self):
             return stream
         return None
 
+    @property
+    def doc(self) -> Optional[str]:
+        """The module docstring."""
```
Comment (cdce8p) [inline astroid/nodes/scoped_nodes/scoped_nodes.py:518]: For full backwards compatibility, we also need a `@doc.setter` method.

**Review by cdce8p — CHANGES_REQUESTED** [review]
This PR contains a lot of different things at the moment which make it difficult to review. I left some early comments, however we should try to get some order into these changes by splitting them up. #1433 doesn't help yet, as it too contains multiple things.

What I have mind
1. Update `Module`, `ClassDef`, `FunctionDef`.
  1.1 The `doc` attribute in `__init__` should be optional with default. Required if we later want to remove it from existing calls.
  1.2 Property and property setter method for `doc` with DeprecationWarning
2. Update other (similar) classes, mainly `object.Property` and `PartialFunction`, with similar changes if necessary.
3. Update code base, to remove the `doc` attribute where it's passed to `__init__` and replace it with `doc_node` to `postinit`. Also update functions which could handle an optional doc_node instead of a string -> `AsStringVisitor`.
4. Add decorator to deprecate `doc` attribute from `__init__` calls.
5. ... Did I miss something 🤔  ... _probably_
6. Update all attribute calls to `.doc` if not already done. `ClassModel.attr___doc__`, `AsStringVisitor._docs_dedent`, ...

**Comment by DanielNoord** [comment]
> This PR contains a lot of different things at the moment which make it difficult to review. I left some early comments, however we should try to get some order into these changes by splitting them up. #1433 doesn't help yet, as it too contains multiple things.
> 
> What I have mind
> 
> 1. Update `Module`, `ClassDef`, `FunctionDef`.
>    1.1 The `doc` attribute in `__init__` should be optional with default. Required if we later want to remove it from existing calls.
>    1.2 Property and property setter method for `doc` with DeprecationWarning
> 2. Update other (similar) classes, mainly `object.Property` and `PartialFunction`, with similar changes if necessary.
> 3. Update code base, to remove the `doc` attribute where it's passed to `__init__` and replace it with `doc_node` to `postinit`. Also update functions which could handle an optional doc_node instead of a string -> `AsStringVisitor`.
> 4. Add decorator to deprecate `doc` attribute from `__init__` calls.
> 5. ... Did I miss something 🤔  ... _probably_

Hmm, that seems like quite a long list. I didn't consider removing `doc` from the `__init__` calls initially and only considered accessing the `doc` attribute.
Although I agree this is something we should do, might it make sense to release `2.11` without the deprecation? That way `pylint` can move on as well.

I'll make a PR that does 1 first.

File: astroid/brain/brain_functools.py (hunk 3 above)
Comment (DanielNoord) [reply astroid/brain/brain_functools.py:104]: Not in their `init` right? They share the `postinit` from `FunctionDef` and don't redefine it, so I don't think we actually need to change those classes? We just need to pass different arguments to their `postinit`.

File: astroid/nodes/as_string.py (hunk 5 above)
Comment (DanielNoord) [reply astroid/nodes/as_string.py:186]: See https://github.com/PyCQA/astroid/pull/1448

**Comment by DanielNoord** [comment]
Closing this in favour of various other PRs.

**Pull request closed without merging**